crates/script/src/broadcast.rs

```diff
@@ -0,0 +1,429 @@
+            // request.
+            let signature =
+                signer.sign_transaction(&tx).await.wrap_err("Failed to sign transaction")?;
+
+            // Submit the raw transaction
+            provider.send_raw_transaction(tx.rlp_signed(&signature)).await?
+        }
+    };
+
+    Ok(pending.tx_hash().to_alloy())
+}
+
+/// How to send a single transaction
+#[derive(Clone)]
+pub enum SendTransactionKind<'a> {
+    Unlocked(Address),
+    Raw(&'a WalletSigner),
+}
+
+/// Represents how to send _all_ transactions
+pub enum SendTransactionsKind {
+    /// Send via `eth_sendTransaction` and rely on the  `from` address being unlocked.
+    Unlocked(HashSet<Address>),
+    /// Send a signed transaction via `eth_sendRawTransaction`
+    Raw(HashMap<Address, WalletSigner>),
+}
+
+impl SendTransactionsKind {
+    /// Returns the [`SendTransactionKind`] for the given address
+    ///
+    /// Returns an error if no matching signer is found or the address is not unlocked
+    pub fn for_sender(&self, addr: &Address) -> Result<SendTransactionKind<'_>> {
+        match self {
+            SendTransactionsKind::Unlocked(unlocked) => {
+                if !unlocked.contains(addr) {
+                    bail!("Sender address {:?} is not unlocked", addr)
+                }
+                Ok(SendTransactionKind::Unlocked(*addr))
+            }
+            SendTransactionsKind::Raw(wallets) => {
+                if let Some(wallet) = wallets.get(addr) {
+                    Ok(SendTransactionKind::Raw(wallet))
+                } else {
+                    bail!("No matching signer for {:?} found", addr)
+                }
+            }
+        }
+    }
+
+    /// How many signers are set
+    pub fn signers_count(&self) -> usize {
+        match self {
+            SendTransactionsKind::Unlocked(addr) => addr.len(),
+            SendTransactionsKind::Raw(signers) => signers.len(),
+        }
+    }
+}
+
+/// State after we have bundled all [TransactionWithMetadata] objects into a single
+/// [ScriptSequenceKind] object containing one or more script sequences.
+pub struct BundledState {
+    pub args: ScriptArgs,
+    pub script_config: ScriptConfig,
+    pub script_wallets: ScriptWallets,
+    pub build_data: LinkedBuildData,
+    pub execution_data: ExecutionData,
+    pub execution_artifacts: ExecutionArtifacts,
+    pub sequence: ScriptSequenceKind,
+}
+
+impl BundledState {
+    pub async fn wait_for_pending(mut self) -> Result<Self> {
+        let futs = self
+            .sequence
+            .sequeneces_mut()
+            .map(|sequence| async move {
+                let rpc_url = sequence.rpc_url();
+                let provider = Arc::new(get_http_provider(rpc_url));
+                receipts::wait_for_pending(provider, sequence).await
+            })
+            .collect::<Vec<_>>();
+
+        let errors =
+            join_all(futs).await.into_iter().filter(|res| res.is_err()).collect::<Vec<_>>();
```

Comment:
we can `filter_map(Result::err)` and format the list using itertools `eyre!("{}", errors.iter().format("\n"))`